s3_scrubber/src/checks.rs

```diff
@@ -178,69 +213,96 @@ pub(crate) struct S3TimelineBlobData {
+        index_part_generation: Generation,
+        s3_layers: HashSet<(LayerFileName, Generation)>,
     },
     Incorrect(Vec<String>),
 }
 
+fn parse_layer_object_name(name: &str) -> Result<(LayerFileName, Generation), String> {
+    match name.rsplit_once('-') {
+        // FIXME: this is gross, just use a regex?
+        Some(gen) if gen.1.len() == 8 => {
+            let layer = gen.0.parse::<LayerFileName>()?;
+            let gen =
+                Generation::parse_suffix(gen.1).ok_or("Malformed generation suffix".to_string())?;
+            Ok((layer, gen))
+        }
+        _ => Ok((name.parse::<LayerFileName>()?, Generation::none())),
+    }
+}
+
 pub(crate) async fn list_timeline_blobs(
     s3_client: &Client,
     id: TenantTimelineId,
     s3_root: &RootTarget,
 ) -> anyhow::Result<S3TimelineBlobData> {
     let mut s3_layers = HashSet::new();
-    let mut index_part_object = None;
-
-    let timeline_dir_target = s3_root.timeline_root(&id);
-    let mut continuation_token = None;
 
     let mut errors = Vec::new();
     let mut keys_to_remove = Vec::new();
 
-    loop {
-        let fetch_response =
-            list_objects_with_retries(s3_client, &timeline_dir_target, continuation_token.clone())
-                .await?;
+    let mut timeline_dir_target = s3_root.timeline_root(&id);
+    timeline_dir_target.delimiter = String::new();
 
-        let subdirectories = fetch_response.common_prefixes().unwrap_or_default();
-        if !subdirectories.is_empty() {
-            errors.push(format!(
-                "S3 list response should not contain any subdirectories, but got {subdirectories:?}"
-            ));
-        }
+    let mut index_parts: Vec<ObjectIdentifier> = Vec::new();
 
-        for (object, key) in fetch_response
-            .contents()
-            .unwrap_or_default()
-            .iter()
-            .filter_map(|object| Some((object, object.key()?)))
-        {
-            let blob_name = key.strip_prefix(&timeline_dir_target.prefix_in_bucket);
-            match blob_name {
-                Some("index_part.json") => index_part_object = Some(object.clone()),
-                Some(maybe_layer_name) => match maybe_layer_name.parse::<LayerFileName>() {
-                    Ok(new_layer) => {
-                        s3_layers.insert(new_layer);
-                    }
-                    Err(e) => {
-                        errors.push(
-                            format!("S3 list response got an object with key {key} that is not a layer name: {e}"),
-                        );
-                        keys_to_remove.push(key.to_string());
-                    }
-                },
-                None => {
-                    errors.push(format!("S3 list response got an object with odd key {key}"));
+    let stream = stream_listing(s3_client, &timeline_dir_target);
+    pin_mut!(stream);
+    while let Some(obj) = stream.next().await {
+        let obj = obj?;
+        let key = match obj.key() {
```

Comment:
this can be a let else

Reply:
Not sure we really have a style preference for let/else vs match in this case.